### PR TITLE
ARC-1969 - Adding jiraHost for fetching logs

### PR DESCRIPTION
**Pull request by krazziekay**
- Added `clientKey` for TenantType and inserted `clientKey` after installation.
- Added a middleware which uses `cookies` for storing the jwt token from Jira side for identifying the Jira instance.
- Used `jwt-decode` to decode the jwt token from Jira and map the corresponding host to display the corresponding logs.

**Video**
https://www.loom.com/share/94a11c7198d54ed9afc48820cf039d98

### CODE REVIEW

File: src/middlewares/jwtMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * This middleware saves the JWT token from Jira into the browser cookies
+ */
+export const jwtTokenMiddleware = (req, res, next) => {
+    // Check the cookies first
+    if (req.cookies.jwt) {
+        res.locals.jwt = req.cookies.jwt;
+    } else {
+        // Check if its in query parameters
+        if (req.query.jwt) {
```
Comment (rachellerathbone) [inline src/middlewares/jwtMiddleware.ts:19]: Why would the jwt be in the query params? Is it encrypted?

File: src/routes/webhooks.ts [hunk 2]
```diff
@@ -1,34 +1,42 @@
...
-            if (data && data.logs.length) {
-                res.render('webhooks.mst', {
-                    index: 'Webhooks Page',
-                    logs: data.logs.sort().reverse()
-                });
-            } else {
+        findOneInDb(
+            { clientKey },
+            (data) => {
+                if (data && data.logs.length) {
```
Comment (rachellerathbone) [inline src/routes/webhooks.ts:15]: Can we use optional chaining here?

File: src/middlewares/jwtMiddleware.ts (hunk 1 above)
Comment (krazziekay) [reply src/middlewares/jwtMiddleware.ts:19]: The authentication is defined as `jwt` within the `atlassian-connect.json`, so when the user goes the post-install page, `jwt` is passed along within the query params.

File: src/routes/webhooks.ts (hunk 2 above)
Comment (krazziekay) [reply src/routes/webhooks.ts]: Updated

File: src/middlewares/jwtMiddleware.ts (hunk 1 above)
Comment (mboudreau) [reply src/middlewares/jwtMiddleware.ts:19]: the jwt token is in the query when it's within the iframe (look it up for g4j)

File: src/middlewares/jwtMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    httpOnly: true
+};
+
+/**
+ * This middleware saves the JWT token from Jira into the browser cookies
+ */
+export const jwtTokenMiddleware = (req, res, next) => {
+    // Check the cookies first
+    if (req.cookies.jwt) {
+        res.locals.jwt = req.cookies.jwt;
```
Comment (mboudreau) [inline src/middlewares/jwtMiddleware.ts:16]: Why are we storing this in the cookie at all if we always have a new jwt token in every iframe page?  G4j uses cookies only for the breakout mechanism

Comment (mboudreau) [reply src/middlewares/jwtMiddleware.ts:16]: Oh, and furthermore, for the breakout mechanism, we store the jwt token "on the way out"

File: src/middlewares/jwtMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    } else {
+        // Check if its in query parameters
+        if (req.query.jwt) {
+            res.cookie('jwt', req.query.jwt, cookieOptions);
+            res.locals.jwt = req.query.jwt;
```
Comment (mboudreau) [inline src/middlewares/jwtMiddleware.ts:21]: This is where we should verify the JWT token using `atlassian-jwt`

File: src/routes/webhooks.ts [hunk 5]
```diff
@@ -1,34 +1,42 @@
 import { Router } from 'express';
 import { findOneInDb, updateToDb} from '../db';
+import {decodeJwtToken} from '../utils';
 
 export const WebhooksRouter = Router();
 
-WebhooksRouter.get('/', (_, res) => {
-    // TODO: Figure out a way to get the host
-    const host = 'https://kmaharjan4.atlassian.net';
+WebhooksRouter.get('/', (_req, res) => {
```
Comment (mboudreau) [inline src/routes/webhooks.ts:7]: Not sure I understand why the webhooks has a GET?

File: src/routes/webhooks.ts [hunk 6]
```diff
@@ -1,34 +1,42 @@
...
-        (data) => {
-            if (data && data.logs.length) {
-                res.render('webhooks.mst', {
-                    index: 'Webhooks Page',
-                    logs: data.logs.sort().reverse()
-                });
-            } else {
+        findOneInDb(
+            { clientKey },
+            (data) => {
```
Comment (mboudreau) [inline src/routes/webhooks.ts:14]: Why are we using callbacks for our DB stuff? we should use async/await for everything

File: src/routes/webhooks.ts [hunk 7]
```diff
@@ -1,34 +1,42 @@
...
-                logs: [ "No logs yet!" ]
-            });
-        }
-    );
+        );
+    } else {
+        res.render('webhooks.mst', {
+            index: 'Webhooks Page',
+            logs: [ "No logs yet!" ]
+        });
```
Comment (mboudreau) [inline src/routes/webhooks.ts:38]: Do you see a problem with this? we have 4 different calls to render, 3 of which are exactly the same.  I think we can reduce this to a single render call :)

File: src/middlewares/jwtMiddleware.ts (hunk 3 above)
Comment (krazziekay) [reply src/middlewares/jwtMiddleware.ts]: Removed the usage of cookies.

File: src/routes/webhooks.ts (hunk 5 above)
Comment (krazziekay) [reply src/routes/webhooks.ts]: Moved out to a separate route.

File: src/routes/webhooks.ts (hunk 7 above)
Comment (krazziekay) [reply src/routes/webhooks.ts]: This has been updated

File: src/routes/webhooks.ts (hunk 6 above)
Comment (krazziekay) [reply src/routes/webhooks.ts]: Updated the db methods using async-await.

File: src/middlewares/jwtMiddleware.ts (hunk 4 above)
Comment (krazziekay) [reply src/middlewares/jwtMiddleware.ts]: Updated `jwt-decode` by `atlassian-jwtf`

File: src/db/index.ts [hunk 8]
```diff
@@ -3,58 +3,60 @@ import path from 'path';
...
             console.log("Found successfully", data);
-            successCallback && successCallback(data);
+            resolve(data);
         }
     });
-};
+});
 
-export const insertToDb = (data: any, successCallback: Function, failureCallback: Function) => {
+export const insertToDb = (data: any) => new Promise((resolve) => {
```
Comment (mboudreau) [inline src/db/index.ts:28]: Hm, I think there's a better way of doing this.  Could we try using https://www.npmjs.com/package/nedb-promises instead?  Works with async/await and has security vulns fixed.

File: src/routes/logs.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        res.render('webhooksLogs.mst', {
+            index: 'Webhooks Page',
+            logs: logsData.logs.sort().reverse()
+        });
+    } else {
+        res.render('webhooksLogs.mst', {
+            index: 'Webhooks Page',
+            logs: ["No logs yet!"]
+        });
+    }
```
Comment (mboudreau) [inline src/routes/logs.ts:22]: I'd rather we just have one render and have the display logic of "no logs yet" be in the template.  Essentially, it either sends an empty array or an array with the logs

File: src/db/index.ts (hunk 8 above)
Comment (krazziekay) [reply src/db/index.ts]: Updated the Promise based function using `nedb-promises`.

File: src/routes/logs.ts (hunk 9 above)
Comment (krazziekay) [reply src/routes/logs.ts]: Moved the empty array check logic in the view.

**Review by mboudreau — APPROVED** [review]

**Pull request merged**